### PR TITLE
Issue 94 - Extending builtin HTMLElements

### PR DESCRIPTION
**Pull request by dannystey**
- added extends configuration to extend the behaviors of builtin html elements such as input, buttons and so on

i got the extends configuration from:
https://github.com/WebReflection/hyperHTML-Element#new-in-v18

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -14,6 +14,9 @@ export default abstract class Component<TProps, TState> extends HyperHTMLElement
   // overwrite if you have a minifier/uglifier in your project
   public static componentName: string;
 
+  // sets the HTMLElement, that should be extended
+  public static extends : string = null;
```
[the remark this answers is not in the document]
Comment (tiagomapmarques) [reply src/index.ts]: I just think it'll be strange to write:
```javascript
class MyComponent extends Component {
  extends: 'HTMLButtonElement';
  ...
}
```

[the remark this answers is not in the document]
Comment (dannystey) [reply src/index.ts]: i changed it to `basedOn`. But we could also change it to anything that fits better. Maybe we have to discuss it with others?!
```js
class MyComponent extends Component {
    basedOn: 'button';
}
```

File: src/register.ts [hunk 2]
```diff
@@ -18,6 +18,12 @@ export const register = (component: typeof Component) => {
     console.warn(`Attempt to re-registering component "${dashedName}".`);
     return;
   }
+
+  let additionalConfiguration: Object = {};
```
[the remark this answers is not in the document]
Comment (dannystey) [reply src/register.ts]: Thats true. Thanks. changed it

[the remark this answers is not in the document]
Comment (dannystey) [reply src/register.ts]: That's true. Thanks. Changed it.

**Comment by dannystey** [comment]
@tiagomapmarques added test cases, too.

File: tests/extendHTMLElement.spec.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      public static basedOn = 'button';
+    }
+
+    TestButton.register();
+
+    const element = new (customElements.get('test-button'))();
+    expect(element).toBeInstanceOf(customElements.get('button'));
+  });
+
+  it('extends a button element and check against input field', (): void => {
```
Comment (tiagomapmarques) [inline tests/extendHTMLElement.spec.ts]: If `a=1`, we can test if `a` is `1` (this is done in L45). This test (L58) is testing that, under the same conditions of L45, `a` is not `2`. Mathematically, I don't think this test is useful (i.e. if `1` is not `2`, and `a=1`, by default, `a!=2`).

IMO would be better to test:
set `a=1` => test if `a` is `1`
set `a=2` => test if `a` is `2`

(`a`= Component, `1`= button, `2` = input)

Comment (dannystey) [reply tests/extendHTMLElement.spec.ts]: I made that test to be sure that it does not return true on every type of element...

Comment (tiagomapmarques) [reply tests/extendHTMLElement.spec.ts]: Yep, I inferred the intention. I'm just saying the end result (the test on the PR) is something that will always return true if we don't hardcode "button" as the `baseOn` attribute.
The test I propose would do the exact same thing, without the mathematical fallacy.

File: tests/extendHTMLElement.spec.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+describe('#Extending HTMLElements', (): void => {
+  // register an extended set on HTMLElements
+  beforeEach((): void => {
+    // eslint-disable-next-line global-require
+    const basicHTML = require('basichtml');
+    // eslint-disable-next-line global-require
+    const basicHTMLElements = require('basichtml-elements');
+    const customElements = new basicHTML.CustomElementRegistry();
+
+    // define things
```
[the remark this answers is not in the document]
Comment (dannystey) [reply tests/extendHTMLElement.spec.ts]: Sorry

File: tests/extendHTMLElement.spec.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
+import Component from '../src';
+
+describe('#Extending HTMLElements', (): void => {
+  // register an extended set on HTMLElements
+  beforeEach((): void => {
+    // eslint-disable-next-line global-require
+    const basicHTML = require('basichtml');
+    // eslint-disable-next-line global-require
+    const basicHTMLElements = require('basichtml-elements');
```
[the remark this answers is not in the document]
Comment (dannystey) [reply tests/extendHTMLElement.spec.ts]: changed it

File: tests/extendHTMLElement.spec.ts (hunk 3 above)
Comment (dannystey) [reply tests/extendHTMLElement.spec.ts]: changed it the way you suggest it.

**Comment by SheepFromHeaven** [comment]
Fixes #94

**Pull request merged**